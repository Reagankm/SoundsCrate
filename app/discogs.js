/* @flow */
import { secureFetch } from './lib/oauth';
import conf from './conf';
import cache from './cache';

import type {
  Folder,
  RecordsPage,
  FolderPage,
  FolderPageReleases
} from './discogsTypes';

export async function getCollectionFolder(folderId: string, pageNumber: number): Promise<RecordsPage> {
  const cachedFolder = cache.get(folderId, pageNumber);
  if (cachedFolder) {
    const { totalPages, records } = cachedFolder;
    return { totalPages, records };
  } else {
     const { totalPages, releases } = await getCollectionFolderPage(folderId, pageNumber);
     const records = releases.map(getReleaseCoverImage);
     const entry = { pageNumber, records };
     cache.update({
       folderId,
       entry,
       totalPages
     });

     return { totalPages, records };
  }
}

export async function getThumbsInFolder(folderId: string, numberOfThumbs: number = 4): Promise<*> {
  const { totalPages, releases } = await getCollectionFolderPage(folderId, 1, numberOfThumbs);
  return releases.map(getReleaseCoverImage);
}

export async function getCollectionFolders(): Promise<Array<Folder>> {
  const {
    discogs: {
      api_url: apiUrl,
      endpoints,
    }
  } = conf;
  const { username } = await getIdentity();
  const foldersCollectionUrl = endpoints.folders.replace('{username}', username);
  const foldersUrl = `${apiUrl}${foldersCollectionUrl}`;
  const foldersResponse = await secureFetch(foldersUrl);
  const { folders } = await getJson(foldersResponse);

  return folders.map(folder => ({
    id: folder.id,
    name: folder.name,
    count: folder.count,
  }));
}

async function getCollectionFolderPage(
  folderId: string,
  pageNumber: number,
  recordsPerPage?: number): Promise<FolderPage> {
  const {
    discogs: {
      api_url: apiUrl,
      endpoints,
      records_per_page,
    }
  } = conf;
  const { username } = await getIdentity();
  const perPage = recordsPerPage || records_per_page;
  const userFolderUrl = endpoints.folder.replace('{username}', username).replace('{id}', folderId.toString());
  const folderUrl = `${apiUrl}${userFolderUrl}?sort=artist&per_page=${perPage}&page=${pageNumber}`;
  const folderResponse = await secureFetch(folderUrl);
  const {
    releases,
    pagination: { pages: totalPages }
  } = await getJson(folderResponse);

  return { totalPages, releases };
}

async function getIdentity(): Promise<*> {
  const {
    discogs: {
      api_url: apiUrl,
      endpoints,
    }
  } = conf;
  const response = await secureFetch(`${apiUrl}${endpoints.identity}`);
  return await getJson(response);
}

function getJson(response: Response): Promise<*> {
  return response.json();
}

function getReleaseCoverImage(release: FolderPageReleases): string {
  return release.basic_information.cover_image;
}
