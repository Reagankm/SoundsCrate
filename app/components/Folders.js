/* @flow */
import React, { Component } from 'react';
import {
  Dimensions,
  Image,
  ScrollView,
  StyleSheet,
  Text,
  TouchableHighlight,
  View,
} from 'react-native';
import { getCollectionFolders, getThumbsInFolder } from '../discogs';
import LoadingIndicator from './LoadingIndicator';

type State = {
  folders: []
};

export default class Folders extends Component {
  state: State = { folders: [] }

  componentWillMount() {
    this.getFolders();
  }

  render(): React.Element<*> {
    const { folders } = this.state;
    const isLoading = !Boolean(folders.length);
    const contentContainerStyle = isLoading ? styles.root : styles.noFlex;
    return (
      <ScrollView style={styles.root} contentContainerStyle={contentContainerStyle}>
        {isLoading && <LoadingIndicator/>}
        {!isLoading && folders.map((folder: { id: string, name: string }, index: number) => (
          <FolderButton
            key={index}
            folder={folder}
            onPress={() => this.openFolder(folder.id)}/>
        ))}
      </ScrollView>
    )
  }

  openFolder(id: string) {
    const { navigate } = this.props.navigation;
    navigate('Folder', { id });
  }

  async getFolders(): Promise<*> {
    const folders = await getCollectionFolders();
    this.setState({ folders });
  }
}

type FolderButtonProps = {
  folder: { id: string, name: string },
  onPress: () => void
};

class FolderButton extends Component {
  props: FolderButtonProps;
  state: { thumbnails: Array<string> } = { thumbnails: [] };

  componentWillMount() {
    this.getFolderThumbs();
  }

  render(): React.Element<*> {
    const { folder } = this.props;
    return (
      <View style={styles.folderContainer}>
        <View style={styles.thumbsContainer}>
          {this.state.thumbnails.map((thumb, index) => (
            <Image
              key={index}
              style={styles.folderThumb}
              source={{ url: thumb }} />
          ))}
        </View>
        <TouchableHighlight
          style={styles.folderTouchable}
          underlayColor={'#ccc'}
          onPress={() => this.props.onPress(folder.id)}>
          <Text style={styles.folderText}>{folder.name}</Text>
        </TouchableHighlight>
      </View>
    );
  }

  async getFolderThumbs(): Promise<*> {
    const { id } = this.props.folder;
    const thumbs = await getThumbsInFolder(id);
    this.setState({ thumbnails: thumbs });
  }
}

const styles = StyleSheet.create({
  root: {
    flex: 1,
  },
  noFlex: {
    flex: 0,
  },
  folderContainer: {
    flex: 1,
    borderBottomWidth: 1,
    borderBottomColor: '#ccc',
    height: 100,
  },
  thumbsContainer: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
    flexWrap: 'wrap',
    opacity: 0.2,
  },
  folderThumb: {
    flex: 1,
    width: 50,
    height: 100,
  },
  folderTouchable: {
    alignItems: 'center',
    justifyContent: 'center',
    position: 'absolute',
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
    zIndex: 1,
  },
  folderText: {
    fontSize: 25,
    fontWeight: 'bold',
  },
});
