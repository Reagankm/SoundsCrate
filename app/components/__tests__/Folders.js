/* @flow */
import React from 'react';
import renderer from 'react-test-renderer';
import Folders from '../Folders';
import { getCollectionFolders } from '../../discogs';

jest.mock('../../discogs', () => ({
  __esModule: true,
  getCollectionFolders: jest.fn(),
}));
jest.mock('../LoadingIndicator', () => 'LoadingIndicator');
jest.mock('TouchableHighlight', () => 'TouchableHighlight');

describe('app/components/Folders', () => {
  it('should render a loader if there is no data', async () => {
    // $FlowFixMe
    getCollectionFolders.mockImplementationOnce(() => Promise.resolve([]));
    const tree = await renderer.create(<Folders />);
    expect(tree).toMatchSnapshot();
  });
  it('should render a list of TouchableHighlights if there is data', async () => {
    // $FlowFixMe
    getCollectionFolders.mockImplementationOnce(() => Promise.resolve([
      {id: 1, name: 'foo'},
      {id: 2, name: 'bar'},
    ]));
    const tree = await renderer.create(<Folders />);
    expect(tree).toMatchSnapshot();
  });
});
