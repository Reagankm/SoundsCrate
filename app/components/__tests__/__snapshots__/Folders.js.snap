// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/components/Folders should render a list of TouchableHighlights if there is data 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flex": 0,
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <FolderButtonContainer
      folder={
        Object {
          "id": 1,
          "name": "foo",
        }
      }
    />
    <FolderButtonContainer
      folder={
        Object {
          "id": 2,
          "name": "bar",
        }
      }
    />
  </View>
</RCTScrollView>
`;

exports[`app/components/Folders should render a loader if there is no data 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flex": 1,
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <LoadingIndicator />
  </View>
</RCTScrollView>
`;
